internal/bot/chat.go

```diff
@@ -0,0 +1,84 @@
+package bot
+
+import (
+	"context"
+	"fmt"
+	"github.com/austinvalle/hammy/internal/llm"
+	"github.com/bwmarrin/discordgo"
+	"log/slog"
+	"math/rand"
+	"slices"
+	"strings"
+)
+
+// todo: implement
+type chatCommand struct {
+	logger *slog.Logger
+	llm    *llm.LLM
+}
+
+func newChatCommand(logger *slog.Logger, llm *llm.LLM) *chatCommand {
+	return &chatCommand{
+		logger: logger,
+		llm:    llm,
+	}
+}
+func (c *chatCommand) Name() string { return "chat" }
+func (c *chatCommand) CanActivate(s *discordgo.Session, m discordgo.Message) bool {
+	safeChannels := []string{"real-talk", "admin", "reaction-roles", "trip-planning-war-room"}
+	channel, err := s.State.Channel(m.ChannelID)
+	if err != nil {
+		c.logger.Error("could not get channel details from state", err)
+		return false
+	}
+	if slices.Contains(safeChannels, channel.Name) {
+		c.logger.Debug("Message received in safe channel, ignoring chat")
+		return false
+	}
+
+	if slices.ContainsFunc(m.Mentions, func(user *discordgo.User) bool {
+		return user.Username == s.State.User.Username
+	}) {
+		return true
+	}
+
+	//if he was mentioned but not tagged
+	if strings.Contains(m.Content, "hammy") {
+		return true
+	}
+	//did we respond in the last 5 messages? if so we should keep talking
+	msgs, err := s.ChannelMessages(m.ChannelID, 5, "", "", m.ID)
+	if err != nil {
+		c.logger.Error("could not get messages from channel", err)
+		return false
+	}
+	if slices.ContainsFunc(msgs, func(msg *discordgo.Message) bool {
+		return msg.Author.ID == s.State.User.ID
+	}) {
+		return true
+	}
+
+	//otherwise 10% chance of responding
+	return rand.Intn(10) == 0
+}
+
+func (c *chatCommand) Handler(ctx context.Context, s *discordgo.Session, m *discordgo.MessageCreate) (*discordgo.MessageSend, error) {
+	//get history
+	//hand to llm
+	//respond
+	msgs, err := s.ChannelMessages(m.ChannelID, 50, "", "", m.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.llm.Chat(ctx, msgs)
+	if err != nil {
+		return nil, err
+	}
+	prefix := fmt.Sprintf("%s#%s said:", s.State.User.Username, s.State.User.Discriminator)
+	resp, _ = strings.CutPrefix(resp, prefix)
+
+	return &discordgo.MessageSend{
+		Content: resp,
+	}, nil
```

Comment:
eventual todo: We should add some trace logs here (or maybe in the can activate) so we can know if he's actively trying to respond to a message